File: src/astronomy/ayanamsa.ts
import { PanchangError } from '../types/errors';
import type { AyanamsaType } from '../types/options';

/**
 * Compute the ayanamsa (precession offset) for a given UTC date.
 *
 * The ayanamsa is subtracted from the tropical (ecliptic) longitude to obtain
 * the sidereal longitude used in Vedic astrology.
 *
 * @param date UTC date to evaluate.
 * @param type Ayanamsa system: `'lahiri'` (default), `'raman'`, or `'krishnamurti'`.
 * @returns    Ayanamsa value in degrees. Typical range ~23–24° for dates near J2000.
 * @throws     `PanchangError` (INVALID_AYANAMSA) for unrecognised type strings.
 *
 * @example
 * ```typescript
 * import { getAyanamsa } from 'panchang-ts';
 * getAyanamsa(new Date('2025-01-01T00:00:00Z'), 'lahiri'); // ~24.10
 * ```
 */
export function computeAyanamsa(date: Date, type: AyanamsaType = 'lahiri'): number {
  const T = julianCenturiesFromJ2000(date);

  switch (type) {
    case 'lahiri':
      return lahiriAyanamsa(T);
    case 'raman':
      return ramanAyanamsa(T);
    case 'krishnamurti':
      return kpAyanamsa(T);
    case 'true-chitra':
      return trueChitraAyanamsa(T);
    case 'thirukanitham':
      return thirukanithamAyanamsa(T);
    default:
      throw new PanchangError(`Unknown ayanamsa type: ${type}`, 'INVALID_AYANAMSA');
  }
}

/**
 * Lahiri (Chitrapaksha) ayanamsa at J2000.0 — **23° 51′ 49.68″**.
 *
 * This is the value DrikPanchang uses, and it is the anchor for every sidereal
 * output in this library, so it is worth recording how it was arrived at.
 *
 * It used to be 23.853211° (23° 51′ 11.6″). That figure is widely repeated for
 * Lahiri, but it is not what Drik — or Swiss Ephemeris, which Drik follows —
 * actually computes, and the 38″ gap was the single largest source of
 * disagreement in the library: it pushed every nakshatra transition ~69 s early
 * and every yoga transition ~129 s early, because those carry the ayanamsa once
 * and twice respectively. (Tithi and karana were unaffected, since Moon − Sun
 * cancels the ayanamsa entirely — that asymmetry is what made the offset
 * visible in `tests/validation/element-endtime-audit.test.ts`.)
 *
 * The replacement is measured, not adopted from a source. Drik publishes its
 * ayanamsa to six decimals on every day-panchang page; solving this constant
 * from nine of them spanning 1950–2050 gives:
 *
 *   1950-01-01  23.86380142      2020-01-01  23.86380151
 *   1975-01-01  23.86380107      2025-01-14  23.86380105
 *   2000-01-01  23.86380112      2030-01-01  23.86380082
 *   2010-01-01  23.86380078      2050-01-01  23.86380114
 *
 * — nine independent determinations agreeing to 0.0096″, over a century-wide
 * baseline. The precession polynomial below was already exactly right: with
 * this constant, the library reproduces Drik's published ayanamsa to ~0.01″
 * across that whole span, so the two are the same function and not merely
 * pinned at one epoch.
 */
const LAHIRI_J2000_DEG = 23.863801;

/**
 * IAU general precession in longitude, arcsec, T in Julian centuries from
 * J2000. Shared by every system below except Raman, which carries its own
 * linear rate by tradition.
 */
function precessionArcsec(T: number): number {
  return 5029.0966 * T + 1.112 * T * T - 0.000006 * T * T * T;
}

/**
 * Offsets from {@link LAHIRI_J2000_DEG} at J2000.0, in degrees.
 *
 * These are the relationships the library has always encoded — previously
 * written out as absolute constants that each embedded the old Lahiri base.
 * Expressing them as offsets means correcting that base carries them along
 * instead of silently changing how each system relates to Lahiri.
 *
 * Only Lahiri is measured against Drik; Drik publishes no value for the others.
 * Preserving their prior offsets is therefore the minimum-assumption choice,
 * not an independent verification of any of them.
 */
const AYANAMSA_OFFSET_FROM_LAHIRI = {
  /** True Chitrapaksha: actual Spica position rather than the mean reference. */
  trueChitra: -0.0006,
  /** Krishnamurti (KP): Lahiri − 4′ 46.6″. */
  kp: -0.079605,
  /** Thirukanitham: Tamil Vakya tradition, Lahiri + 1′ 6.4″. */
  thirukanitham: +0.018456,
  /** B. V. Raman: Lahiri − 1° 23′ 33.8″ at J2000 (own rate thereafter). */
  raman: -1.392722,
} as const;

/**
 * True Chitrapaksha ayanamsa — anchored to Chitra star (Spica) at 180°
 * sidereal. Differs from Lahiri by ~2 arcsec at J2000 (Lahiri uses a
 * mean-Spica reference; True Chitra uses the actual stellar position).
 */
function trueChitraAyanamsa(T: number): number {
  return LAHIRI_J2000_DEG + AYANAMSA_OFFSET_FROM_LAHIRI.trueChitra + precessionArcsec(T) / 3600;
}

/**
 * Thirukanitham ayanamsa — Tamil-tradition variant used in some South
 * Indian (Vakya) panchangs. ~1.1 arcmin offset from Lahiri at J2000;
 * same precession rate.
 */
function thirukanithamAyanamsa(T: number): number {
  return LAHIRI_J2000_DEG + AYANAMSA_OFFSET_FROM_LAHIRI.thirukanitham + precessionArcsec(T) / 3600;
}

function lahiriAyanamsa(T: number): number {
  return LAHIRI_J2000_DEG + precessionArcsec(T) / 3600;
}

function ramanAyanamsa(T: number): number {
  const annualRateDeg = 50.3304 / 3600;
  return LAHIRI_J2000_DEG + AYANAMSA_OFFSET_FROM_LAHIRI.raman + annualRateDeg * T * 100;
}

function kpAyanamsa(T: number): number {
  return LAHIRI_J2000_DEG + AYANAMSA_OFFSET_FROM_LAHIRI.kp + precessionArcsec(T) / 3600;
}

function julianCenturiesFromJ2000(date: Date): number {
  const jd = dateToJulianDay(date);
  return (jd - 2451545.0) / 36525.0;
}

export function dateToJulianDay(date: Date): number {
  return date.getTime() / 86_400_000 + 2440587.5;
}

File: source/go/internal/astronomy/ayanamsa.go
package astronomy

import (
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
)

func ComputeAyanamsa(ms int64, typ types.AyanamsaType) (float64, error) {
	t := julianCenturiesFromJ2000(ms)

	switch typ {
	case types.Lahiri:
		return lahiriAyanamsa(t), nil
	case types.Raman:
		return ramanAyanamsa(t), nil
	case types.Krishnamurti:
		return kpAyanamsa(t), nil
	case types.TrueChitra:
		return trueChitraAyanamsa(t), nil
	case types.Thirukanitham:
		return thirukanithamAyanamsa(t), nil
	default:
		return 0, types.Codef(types.ErrInvalidAyanamsa, "Unknown ayanamsa type: %s", typ)
	}
}

// Solved from Drik, not the widely repeated 23.853211°; typed so the offset sums do not constant-fold.
const lahiriJ2000Deg float64 = 23.863801

// IAU general precession in longitude, arcsec; the casts are FMA barriers.
func precessionArcsec(t float64) float64 {
	return float64(5029.0966*t) + float64(1.112*t*t) - float64(0.000006*t*t*t)
}

const (
	offsetTrueChitra    float64 = -0.0006
	offsetKP            float64 = -0.079605
	offsetThirukanitham float64 = +0.018456
	// Measured against SwissEph; the literature value −1.392722 is wrong.
	offsetRaman float64 = -1.453010
)

func trueChitraAyanamsa(t float64) float64 {
	return lahiriJ2000Deg + offsetTrueChitra + precessionArcsec(t)/3600
}

func thirukanithamAyanamsa(t float64) float64 {
	return lahiriJ2000Deg + offsetThirukanitham + precessionArcsec(t)/3600
}

func lahiriAyanamsa(t float64) float64 {
	return lahiriJ2000Deg + precessionArcsec(t)/3600
}

func ramanAyanamsa(t float64) float64 {
	return lahiriJ2000Deg + offsetRaman + precessionArcsec(t)/3600
}

func kpAyanamsa(t float64) float64 {
	return lahiriJ2000Deg + offsetKP + precessionArcsec(t)/3600
}

func julianCenturiesFromJ2000(ms int64) float64 {
	jd := DateToJulianDay(ms)
	return (jd - 2451545.0) / 36525.0
}

// UT, not TT: ΔT is 0.001″ of ayanamsa here.
func DateToJulianDay(ms int64) float64 {
	return float64(ms)/86_400_000 + 2440587.5
}
